src/Akka.Streams.Kafka/Stages/CommittableConsumerStage.cs

```diff
@@ -32,51 +33,56 @@ public override ILogicAndMaterializedValue<Task> CreateLogicAndMaterializedValue
         }
     }
 
-    internal class KafkaCommittableSourceStage<K, V> : TimerGraphStageLogic
+    internal class KafkaCommittableSourceStage<K, V> : GraphStageLogic
     {
         private readonly ConsumerSettings<K, V> _settings;
         private readonly ISubscription _subscription;
-        private readonly Outlet<CommittableMessage<K, V>> _out;
         private IConsumer<K, V> _consumer;
 
-        private Action<ConsumeResult<K, V>> _messagesReceived;
         private Action<IEnumerable<TopicPartition>> _partitionsAssigned;
         private Action<IEnumerable<TopicPartitionOffset>> _partitionsRevoked;
         private readonly Decider _decider;
 
-        private const string TimerKey = "PollTimer";
-
-        private readonly Queue<CommittableMessage<K, V>> _buffer;
         private IEnumerable<TopicPartition> _assignedPartitions;
-        private volatile bool _isPaused;
         private readonly TaskCompletionSource<NotUsed> _completion;
+        private readonly CancellationTokenSource _cancellationTokenSource;
 
         public KafkaCommittableSourceStage(CommittableConsumerStage<K, V> stage, Attributes attributes, TaskCompletionSource<NotUsed> completion) : base(stage.Shape)
         {
             _settings = stage.Settings;
             _subscription = stage.Subscription;
-            _out = stage.Out;
             _completion = completion;
-            _buffer = new Queue<CommittableMessage<K, V>>(stage.Settings.BufferSize);
+            _cancellationTokenSource = new CancellationTokenSource();
 
             var supervisionStrategy = attributes.GetAttribute<ActorAttributes.SupervisionStrategy>(null);
             _decider = supervisionStrategy != null ? supervisionStrategy.Decider : Deciders.ResumingDecider;
 
-            SetHandler(_out, onPull: () =>
+            SetHandler(stage.Out, onPull: () =>
             {
-                if (_buffer.Count > 0)
-                {
-                    Push(_out, _buffer.Dequeue());
-                }
-                else
+                try
                 {
-                    if (_isPaused)
+                    var message = _consumer.Consume(_cancellationTokenSource.Token);
+                    if (message == null) // No message received, or consume error occured
+                        return;
+
+                    var consumer = _consumer;
+                    var commitableOffset = new CommitableOffset(
```

Comment:
Does the `groupId` string need to be hard coded?

Reply:
And if so, why? Asking because I don't know enough about the Kafka driver to fully understand myself.

Reply:
Indeed, did not notice this, left as it was... Thanks for noticing! GroupId is the thing we may pass to consumer factory and is constant for all messages we consume - it is stored in _settings stage field, so I will pass real value instead of hard-coded value in separate PR.